### ISSUE TITLE
☂️ Multi View for Windows/MacOS

### ISSUE THREAD

**Issue by goderbauer** [flutter/flutter#142845]
## Add basic multi-view rendering support

- [x] Design documents
    - [x] https://flutter.dev/go/multiple-views
    - [x] https://flutter.dev/go/multi-view-embedder-apis
    - [x] https://flutter.dev/go/desktop-multi-view-runner-apis
    - [x] https://flutter.dev/go/multi-view-sync-over-async
- [x] Add multi-view `dart:ui` APIs
    - [x] https://github.com/flutter/flutter/issues/112205
- [x] Resolve framework's warm-up frame problem
  - [x] https://github.com/flutter/engine/pull/50570
  - [x] https://github.com/flutter/flutter/pull/143290
- [x] Re-land multi-view rasterization pipeline: https://github.com/flutter/engine/pull/51186
    - [x] https://github.com/flutter/flutter/issues/144444
    - [x] https://github.com/flutter/flutter/issues/144445
    - [x] https://github.com/flutter/flutter/issues/144584
- [x] Add multi-view rendering embedder APIs
  - [x] send window metrics with view IDs: https://github.com/flutter/engine/pull/46174
  - [x] https://github.com/flutter/engine/pull/46213
- [ ] #144810 (rendering only)
  - [x] Windows: https://github.com/flutter/engine/pull/51923
  - [ ] macOS
- [ ] Add new multi-view runner APIs (as "pubternal")
  - [ ] https://github.com/flutter/flutter/issues/143767
  - [ ] https://github.com/flutter/flutter/issues/145874
- [x] Make Windows embedder multi-view ready
  - [x] https://flutter.dev/go/windows-multi-view-ownership-updates 
  - [x] https://github.com/flutter/flutter/issues/137267
  - [x] https://github.com/flutter/flutter/issues/143765

## Work after rendering

- Make all built-in features multi-view compatible
  - [ ] https://github.com/flutter/flutter/issues/144806 
  - [ ] Built-in plugins: text input, cursor, app lifecycle, etc.
  - [ ] Accessibility
      - [ ] https://github.com/flutter/flutter/issues/112207
- [ ] Make all plugins multi-view compatible
  - [ ] Decouple plugin registration from view initialization
    - [ ] Ensure plugins do not crash if there is no view
  - [ ] Make runner APIs public
  - [ ] Ensure plugins do not use runner APIs that are incompatible with multiple views
- Bugs
  - [x] https://github.com/flutter/flutter/issues/146251
  - [ ] https://github.com/flutter/flutter/issues/135534
  - [ ] Engine should drop pointer events sent to nonexistent views
  - [ ] https://github.com/flutter/flutter/issues/130165
- Next phase
  - [ ] Add Dart windowing APIs
    - [ ] https://github.com/flutter/flutter/issues/30736
  - [ ] Update DevTools
    - [ ] Update frame timings 
  - [ ] Hide new windows until first frame to prevent black flashes
  - [ ] https://github.com/flutter/flutter/issues/144633
    - [ ] https://github.com/flutter/flutter/issues/131651
  - [ ] #146249
  - [ ] https://github.com/flutter/flutter/issues/145712
  - [ ] https://github.com/flutter/flutter/issues/146013
- Clean ups
  - [ ] Deprecate embedder APIs that aren't multi-view compatible
  - [ ] #146248

Other platforms:
- [ ] https://github.com/flutter/flutter/issues/138178
- [ ] https://github.com/flutter/flutter/issues/138168

**Comment by knopp** [comment]
@goderbauer, `VsyncCallback` doesn't seem to contain any viewId, is there going to a separate method for multiview?

**Comment by dkwingsmt** [comment]
@knopp I'm pretty sure the current consensus is that Flutter will use one vsync for all views, therefore `VsyncCallback` shouldn't need a view ID.

**Comment by knopp** [comment]
> @knopp I'm pretty sure the current consensus is that Flutter will use one vsync for all views, therefore `VsyncCallback` shouldn't need a view ID.

How would that work with multiple displays? For example I have routinely 120hz on one display (promotion) and 60hz on the other.

**Comment by loic-sharma** [comment]
@knopp Our plan for the short-term future is to render all views on each frame using a single vsync. This does mean that we'll be rendering views outside of their monitor's desired vsync. This is something we plan to address after the minimum viable product.

**Comment by dkwingsmt** [comment]
Yeah, during the discussion we were aware that it wouldn't handle the case where different windows are shown on different monitors. We deemed it of lower priority because:
1. Many apps don't support cross-monitor display well either.
2. Having different vsync for different monitors might require big changes to Flutter. Most notably, the widget tree needs to be rebuilt at both 60Hz and 144Hz, basically 204 times per second. (We'll probably figure out a way eventually.)

**Comment by goderbauer** [comment]
To ensure a stronger focus on our primary goals (as detailed in our [2024 roadmap](https://github.com/flutter/flutter/wiki/Roadmap)), we unfortunately had to pause work on this. The current status of the feature (including necessary future work items) is documented at https://docs.google.com/document/d/13E27tD8_9f6lDgwg3MpGNTV8XIRCZH3ByI-t9kI9IUM/edit?usp=sharing.

**Comment by GroovinChip** [comment]
> To ensure a stronger focus on our primary goals (as detailed in our [2024 roadmap](https://github.com/flutter/flutter/wiki/Roadmap)), we unfortunately had to pause work on this. The current status of the feature (including necessary future work items) is documented at https://docs.google.com/document/d/13E27tD8_9f6lDgwg3MpGNTV8XIRCZH3ByI-t9kI9IUM/edit?usp=sharing.

Thank you for the update on this feature. I'm disappointed that it is still not a higher priority, but the update is very much appreciated.

**Comment by wanjm** [comment]
I am so sorry to hear that.

**Comment by wanjm** [comment]
I hope Multi-view rendering can be finished, so that we can create windows with channel and use multi windows;

**Comment by kammcs** [comment]
The lack of multiple windows hinders a number of use cases for desktop app development.

This is, for me at least, the most important feature for flutter desktop support.

Hope it gets reprioritized soon.

**Comment by rekire** [comment]
Might sound stupid simple, but what about contributing? Flutter is open source you can add those features when you have the time. A company who really needs that can invest their "man power"* to add the required features.

@goderbauer do you have the capacity to review those changes if someone would contribute?

<sub>*) I'm not aware of a more inclusive term.</sub>

**Comment by wanjm** [comment]
I come here every to find surprise

**Comment by Ansh-Rathod** [comment]
I just this video [https://www.youtube.com/watch?v=vtB-teu57vw](https://www.youtube.com/watch?v=vtB-teu57vw) this feels like superpower! that every window is in same widget tree I was using multi_desktop_window plugin and it's just nightmare to pass data window to window also it's buggy. 

Please resume this issue as soon as possible it's would be super cool to have this feature.

**Comment by s1nistr4** [comment]
We got the GTA 6 trailer before flutter got multi window support

**Comment by HAIDERALI79** [comment]
I will use Compose multi-platform instead its support multi-window

**Comment by pichillilorenzo** [comment]
Hello guys, I released a new plugin: [window_manager_plus](https://pub.dev/packages/window_manager_plus)

This plugin allows Flutter desktop apps to create and manage multiple windows, such as resizing and repositioning, and communicate between them.

This is a fork and a re-work of the original [window_manager](https://pub.dev/packages/window_manager) plugin. With inspiration from the [desktop_multi_window](https://pub.dev/packages/desktop_multi_window) plugin, this new implementation allows the creation and management of multiple windows.

Linux is not currently supported.

I have already used it in my [Flutter Browser App](https://github.com/pichillilorenzo/flutter_browser_app), that is a Full-Featured Mobile and Desktop Browser App (such as the Google Chrome mobile browser) created using Flutter and the features offered by my other plugin, the flutter_inappwebview plugin.

So, with this plugin, I can manage multiple windows on Flutter.

**Comment by Wassiem-Ibrahim** [comment]
It's frustrating to find out that work is paused on this feature. I hope that it will be added to the 2025 roadmap.

**Comment by Wassiem-Ibr** [comment]
> Hello guys, I released a new plugin: [window_manager_plus](https://pub.dev/packages/window_manager_plus)
> 
> This plugin allows Flutter desktop apps to create and manage multiple windows, such as resizing and repositioning, and communicate between them. Also, plugin registration to make plugins work on all Flutter windows. Follow the instructions of the plugin.
> 
> This is a fork and a re-work of the original [window_manager](https://pub.dev/packages/window_manager) plugin. With inspiration from the [desktop_multi_window](https://pub.dev/packages/desktop_multi_window) plugin, this new implementation allows the creation and management of multiple windows.
> 
> Linux is not currently supported.
> 
> I have already used it in my [Flutter Browser App](https://github.com/pichillilorenzo/flutter_browser_app), that is a Full-Featured Mobile and Desktop Browser App (such as the Google Chrome mobile browser) created using Flutter and the features offered by my other plugin, the flutter_inappwebview plugin.
> 
> So, with this plugin, I can manage multiple windows on Flutter.

this package creates a new screen with the same content of the old screen, what is actually needed is a solution to open a new screen with new content

**Comment by pichillilorenzo** [comment]
@Wassiem-Ibr yes but you need to the `args` of the main function and get the ID of the new Window (0 is the first window) or pass also other custom arguments and then you should manage it in your main function.
You can open whatever new Flutter widget tree you want based on the args you passed.

**Comment by pichillilorenzo** [comment]
By the way, there are some problems on macOS about the current multi-window support in general: https://github.com/flutter/flutter/issues/133533

**Comment by Saviq** [comment]
Hey all, we're happy to show the work we've been doing on this here at Ubuntu/Canonical:
- https://github.com/flutter/flutter/pull/157525

The design doc for it is here ([waiting](https://github.com/flutter/website/pull/11315) for a Go link): 
https://docs.google.com/presentation/d/1dvJmFzadGwZZHHeoRtRJOyZ-dfb3fTJCck87JJL7G84/edit

We'll be showing it off at the Ubuntu Summit on Sunday 27th, come say hi if you're around or tune in on the live stream, there should be a way to ask questions, too:
https://events.canonical.com/event/51/contributions/533/

**Comment by GroovinChip** [comment]
> Hey all, we're happy to show the work we've been doing on this here at Ubuntu/Canonical:
> 
> - https://github.com/flutter/flutter/pull/157525
> 
> 
> 
> The design doc for it is here ([waiting](https://github.com/flutter/website/pull/11315) for a Go link): 
> 
> https://docs.google.com/document/d/1eQG-IS7r4_S9_h50MY_hSGwUVtgSTRSLzu7MLzPNd2Y/edit
> 
> 
> 
> We'll be showing it off at the Ubuntu Summit on Sunday 27th, come say hi if you're around or tune in on the live stream, there should be a way to ask questions, too:
> 
> https://events.canonical.com/event/51/contributions/533/

Incredible work! Will this support macOS as well?

**Comment by Saviq** [comment]
> Incredible work! Will this support macOS as well?

Yes, we want to bring this to all platforms, just had to start somewhere :)

It's designed in a way that should allow you write it once, and have it run on all platforms - not only desktop, even.

**Comment by kjxbyz** [comment]
> We'll be showing it off at the Ubuntu Summit on Sunday 27th, come say hi if you're around or tune in on the live stream, there should be a way to ask questions, too:
> 
> https://events.canonical.com/event/51/contributions/533/

@Saviq Due to the time difference, I can't participate immediately. If possible, can you upload the video to YouTube?

**Comment by rekire** [comment]
I just want to mention this link:
https://docs.google.com/presentation/d/1dvJmFzadGwZZHHeoRtRJOyZ-dfb3fTJCck87JJL7G84/edit#slide=id.g30deeab9cfd_2_0
There you can see their demo.

**Comment by Saviq** [comment]
> @Saviq Due to the time difference, I can't participate immediately. If possible, can you upload the video to YouTube?

Hey, we couldn't actually live stream it, but our talk was recorded and will be posted to YouTube after some editing indeed. Will post it when it's here.

**Comment by shujaatak** [comment]
@Saviq Great Work! But the PR is not merged and has been changed to a draft.

**Comment by Chappie74** [comment]
Brought tears to my eyes seeing the demo. Thank you for working on this <3.

**Comment by psovit** [comment]
While the multi view support is still being developed, what do you guys think of this approach:
- Have two different Flutter app projects
- Each project will generate its own exe or similar app output
- One app can trigger open another app using `Process` commands
- Keep data shared across both apps using `Shared Prefs` or other processes
? 
Anyone tried this?

**Comment by damywise** [comment]
> While the multi view support is still being developed, what do you guys think of this approach:
> 
> * Have two different Flutter app projects
> * Each project will generate its own exe or similar app output
> * One app can trigger open another app using `Process` commands
> * Keep data shared across both apps using `Shared Prefs` or other processes
>   ?
>   Anyone tried this?

That's very much almost how all the multi window plugin work.

**Comment by MurtadhaS** [comment]
> While the multi view support is still being developed, what do you guys think of this approach:
> 
>     * Have two different Flutter app projects
> 
>     * Each project will generate its own exe or similar app output
> 
>     * One app can trigger open another app using `Process` commands
> 
>     * Keep data shared across both apps using `Shared Prefs` or other processes
>       ?
>       Anyone tried this?

aren't the canonical PR above enough?
I think it provides what is expected of this feature.

**Comment by Ferry-200** [comment]
> Hey all, we're happy to show the work we've been doing on this here at Ubuntu/Canonical:
> 
> * [example: material 3 multi-window demo #157525](https://github.com/flutter/flutter/pull/157525)
> 
> The design doc for it is here ([waiting](https://github.com/flutter/website/pull/11315) for a Go link): https://docs.google.com/document/d/1eQG-IS7r4_S9_h50MY_hSGwUVtgSTRSLzu7MLzPNd2Y/edit
> 
> We'll be showing it off at the Ubuntu Summit on Sunday 27th, come say hi if you're around ~or tune in on the live stream, there should be a way to ask questions, too~: https://events.canonical.com/event/51/contributions/533/
> 
> EDIT: We couldn't stream it live after all, will reply here when the recording is published.

The recording is published [Ubuntu Summit 2024 | Flutter State of The Union](https://youtu.be/2-GdfzEVq8k?si=cJYtZB8sT-DJn0IC)

**Comment by FMorschel** [comment]
I don't think this should have been closed, @mattkae

**Comment by GroovinChip** [comment]
Wrongly closed once again :|

**Comment by FMorschel** [comment]
Please change your phrasing everyone! Prefer something like: `Fixes part of` so this doesn't get closed multiple times. Thanks 💙!

**Comment by stan-at-work** [comment]
Is it me or does this look like this issues is again beeing worked on actively?

**Comment by RedyAu** [comment]
Canonical has been carrying the torch for the last little while and are planning on completing it, last we've heard 🤞 Their efforts are highly appreciated.

**Comment by nathantaal** [comment]
But they also hoping on the community to help them implement it for Mac, for example.

**Comment by Saviq** [comment]
> But they also hoping on the community to help them implement it for Mac, for example.

We actually found a way to work on this for macOS, too, watch this space ;)

**Comment by stan-at-work** [comment]
> Canonical has been carrying the torch for the last little while and are planning on completing it, last we've heard 🤞 Their efforts are highly appreciated.

They are only actively working on Linux right now and not Windows? Is that correct?

**Comment by nathantaal** [comment]
They did Windows first, Linux came later and there at that stage right now. But it essentially is al Canonical. Check the video linked above if you want to know more 🙂
________________________________

**Comment by stan-at-work** [comment]
> They did Windows first, Linux came later and there at that stage right now. But it essentially is al Canonical. Check the video linked above if you want to know more 🙂
> […](#)

Thanks 😸

**Comment by lingyun000** [comment]
When can we try this on macOS?

**Comment by Saviq** [comment]
> When can we try this on macOS?

You can _try_ this already in https://github.com/flutter/flutter/pull/167393 👍 - it does still require you to [build a custom engine](https://github.com/flutter/flutter/blob/master/engine/src/flutter/docs/contributing/Compiling-the-engine.md), though.

https://github.com/user-attachments/assets/aad2fad7-1ca8-4f51-b67a-3e387cc6913d

**Comment by lingyun000** [comment]
> > When can we try this on macOS?
> 
> You can _try_ this already in [#167393](https://github.com/flutter/flutter/pull/167393) 👍 - it does still require you to [build a custom engine](https://github.com/flutter/flutter/blob/master/engine/src/flutter/docs/contributing/Compiling-the-engine.md), though.
> 
>  multi-window-recording.mov

Could you please provide a pre-compiled SDK and a simple multi-window example?

**Comment by makri-aymen** [comment]
When this feature will be available, in stable release? any information about what is left to be released?

**Comment by wanjm** [comment]
yes, i also want to know the latest status information of this feature;

**Comment by Saviq** [comment]
@wanjm it's all in, behind a `flutter config --enable-windowing` feature flag:

https://github.com/flutter/flutter/tree/master/examples/multiple_windows

We'll do more to make this known over the coming weeks.

**Comment by mauricev** [comment]
It would be nice if bug https://github.com/flutter/flutter/issues/184701 could be addressed before migrating off master.
